flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java

```diff
@@ -0,0 +1,214 @@
+            ClassLoader userCodeClassLoader,
+            CheckpointCoordinator checkpointCoordinator,
+            String targetDirectory,
+            boolean advanceToEndOfEventTime,
+            CompletableFuture<String> operationCompletionFuture) {
+        super(
+                context,
+                executionGraph,
+                executionGraphHandler,
+                operatorCoordinatorHandler,
+                logger,
+                userCodeClassLoader);
+        this.context = context;
+        this.checkpointCoordinator = checkpointCoordinator;
+        this.targetDirectory = targetDirectory;
+        this.advanceToEndOfEventTime = advanceToEndOfEventTime;
+        this.operationCompletionFuture = operationCompletionFuture;
+    }
+
+    @Override
+    public void onEnter() {
+        final ExecutionGraph executionGraph = getExecutionGraph();
+        // we stop the checkpoint coordinator so that we are guaranteed
+        // to have only the data of the synchronous savepoint committed.
+        // in case of failure, and if the job restarts, the coordinator
+        // will be restarted by the CheckpointCoordinatorDeActivator.
+        checkpointCoordinator.stopCheckpointScheduler();
+
+        CompletableFuture<String> savepointFuture =
+                checkpointCoordinator
+                        .triggerSynchronousSavepoint(advanceToEndOfEventTime, targetDirectory)
+                        // handle errors while creating the savepoint
+                        .handleAsync(
+                                this::handleSavepointOperationResult,
+                                context.getMainThreadExecutor())
+                        // Wait for job to be finished
+                        .thenCompose(
+                                path ->
+                                        executionGraph
+                                                .getTerminationFuture()
+                                                .thenApply(
+                                                        status ->
+                                                                new JobStatusAndSavepointPath(
+                                                                        status, path)))
+                        // check that the job finished successfully
+                        .handleAsync(
+                                this::handleStopWithSavepointOperationResult,
+                                context.getMainThreadExecutor());
+
+        FutureUtils.forward(savepointFuture, operationCompletionFuture);
+    }
+
+    private String handleSavepointOperationResult(
+            CompletedCheckpoint completedCheckpoint, Throwable throwable) {
+        if (throwable != null) {
+            context.runIfState(this, () -> handleAnyFailure(throwable));
+            throw new SavepointCreationException(
+                    "Unable to stop with savepoint. Error while creating the savepoint.",
+                    throwable);
```

Comment:
Handling the failure and sending the result to the user happens here in two places. The latter is even obfuscated a bit through throwing an exception which is forwarded to the result future at some other place. This is a rather complicated control flow. Couldn't we do this in the failure handling logic?

Are we treating savepoint failures as global failures? It looks a bit like this. I think this is wrong. A savepoint failure should only fail the operation and transition us back to the `Executing` state.